src/python/pants/engine/target.py

```diff
@@ -328,6 +328,26 @@ def _find_registered_field_subclass(
         )
         return cast(Optional[Type[_F]], subclass)
 
+    @final
+    def maybe_get(self, field: Type[_F]) -> Optional[_F]:
```

Comment:
It would likely be nice to be able to supply a default. That would require some work on the Field API fwict though. How exactly to spell the following needs support:
```python
timeout_field_instance = target.maybe_get(Timeout, Timeout(15))  # raw_value -> hydrate or back door for pre-hydrated known valid?
timeout_field_instance = target.maybe_get(Timeout, 15)  # If Fields were type parameterized by their value, maybe this style could be supported.
```

Reply:
Rather, `Field`s always have default values embedded in their `.value` implementations because the `raw_value` in their constructor may always be `None`, which represents the user not specifying that field explicitly in the BUILD file:

https://github.com/pantsbuild/pants/blob/9259097a86a6ec7294f280a22d2b91053d1b9d31/src/python/pants/engine/target.py#L426-L433

https://github.com/pantsbuild/pants/blob/9259097a86a6ec7294f280a22d2b91053d1b9d31/src/python/pants/backend/python/rules/targets.py#L120-L131

Sometimes, that default is as simple as `None`:

https://github.com/pantsbuild/pants/blob/9259097a86a6ec7294f280a22d2b91053d1b9d31/src/python/pants/engine/target.py#L436-L451

--

So, to avoid duplication of having to define "what does default mean" in two places, we would instead do this:

```
def maybe_get(self, field: Type[_F]) -> _F:
    result = ...
    if result is None:
       return field(raw_value=None, address=self.address)
```

I don't think `maybe_get` is very descriptive, at that point. Instead, I think it would be `Target.get_or_default()`?

What do you think about `Target.get_or_default()`? I really like that! I think it's better than `maybe_get()`. It avoids Stu's concern from this morning of downstream rules having an explosion of how to express if a target has a field

* field is simply not registered (i.e. `Target.maybe_get()`)
* field is registered, but not specified in the BUILD file (i.e. `Field(raw_value=None)`)
* field is registered and specified in the BUILD file (e.g. `Field(raw_value=True)`)

Reply:
I think you miss the point of usage. Forgetting this Target / Field API, I often want to say the spiritual equivalent of:
```python
timeout = target.timeout or self._default_timeout
```

That means I want to pass a default value I define - not that the field type defines - for concise code.

Reply:
>> I often want to say the spiritual equivalent of:
>
> timeout = target.timeout or options.default_timeout

I see. Hm. What about this?

```
tgt.get_or_default(ZipSafe)  # use the default for `ZipSafe`
tgt.get_or_default(ZipSafe, pex_subsystem.options.zip_safe)
```

The reason I like that more than

Reply:
It seems to me get_or_default is only needed to cover the default value of the Field type case.

```python
tgt.get_or_default(ZipSafe)  # use the default for ZipSafe as a fallback.
tgt.get(ZipSafe, pex_subsystem.options.zip_safe)  # use pex_subsystem.options.zip_safe as a fallback value to create a ZipSafe with
tgt.get(ZipSafe)  # no fallbacks - raise if not set
```

Reply:
> It seems to me get_or_default is only needed to cover the default value of the Field type case.

Ah, I love it. Then we have parity with `dict.get()`. Thank you for the suggestions! I'll rewrite this.

Reply:
In fact, if you supported indexing, you could just make this work like a dict:
```python
tgt[ZipSafe]  # no fallbacks - raise if not set
tgt.get(ZipSafe) # use the default for ZipSafe as a fallback.
tgt.get(ZipSafe, pex_subsystem.options.zip_safe)  # use pex_subsystem.options.zip_safe as a fallback value to create a ZipSafe with
```

That said - way too much paint on the shed. Just go for it. We're still in plastic mode.

Reply:
Oh, wait, no. I spoke too soon. Sorry.

Where this falls apart is that `tgt.get` returns a fully typed `Field`. It does not return the hydrated value, by design. This allows us to do things like pass an instance of `SourcesField` to rules like `strip_source_roots`, for example: https://github.com/pantsbuild/pants/pull/9335.

The actual spiritual equivalent you were after is this:

```python
# note the `.value` call
timeout = tgt.get(Timeout).value or options.default_timeout
```

--

We could use the below because we will pass the `raw_value` to the `Field` constructor, so we can still return a `Field`:

```python
zip_safe1 = tgt.get_or_default(ZipSafe, pex_subsystem.options.zip_safe)
zip_safe2 = ZipSafe(pex_subsystem.options.zip_safe, address=tgt.address)
assert zip_safe1 == zip_safe2
```

I don't love the name `Target.get_or_default()` because `or_default` suggests to me that you are going to pass a default `Field` instance in the parameter. But, we can get around this by enforcing kwargs:

```python
def get_or_default(field: Type[_F], *, default_raw_value: Optional[Any] = None) -> _F:
  ...
```

Reply:
> The actual spiritual equivalent you were after is this:

No, I was just hoping for what my initial exmaple showed, getting back field instances by possibly taking a raw value default and constructing a field instance.:
```python
timeout_field_instance = target.maybe_get(Timeout, Timeout(15))  # raw_value -> hydrate or back door for pre-hydrated known valid?
timeout_field_instance = target.maybe_get(Timeout, 15)  # If Fields were type parameterized by their value, maybe this style could be supported.
```